### PR TITLE
Adds CSV export command

### PR DESCRIPTION
**Pull request by tuj**
* Adds a command to export a day's hits/nohits to csv.
* Adds target abstraction to extract statistics function to allow for exporting to csv.
* Adds identifierType to Entry.

### CODE REVIEW

**Comment by cableman** [comment]
It would be nice to have day-range, so the last week could be extracted.

Not sure that i understand how you request that it's no-hits that we want. Where do I set the ES message that i want to extract?

File: src/Command/ExtractStatisticsDaysCommand.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
+<?php
+
+/**
+ * @file
+ * Contains a command to extract statistics from Elasticsearch.
```
Comment (cableman) [inline src/Command/ExtractStatisticsDaysCommand.php]: Technically you have an interface. So it's not guarantied that it is ES it's coming from but StatisticsExtractionService :-).

File: src/Command/ExtractStatisticsDaysCommand.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+ */
+class ExtractStatisticsDaysCommand extends Command
+{
+    private $extractionService;
+    private $dateFormat = 'd-m-Y';
+
+    protected static $defaultName = 'app:extract-days';
+
+    /**
+     * ExtractStatisticsCommand constructor.
```
Comment (cableman) [inline src/Command/ExtractStatisticsDaysCommand.php]: Think the comment should be "ExtractStatisticsDaysCommand" ?

File: src/Command/ExtractStatisticsDaysCommand.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        $this->setDescription('Extracts statistics for a range of days.');
+
+        $this->addArgument(
+            'filename',
+            InputArgument::REQUIRED,
+            'Csv filename to extract to.',
+            null
+        );
+
+        $this->addArgument(
```
Comment (cableman) [inline src/Command/ExtractStatisticsDaysCommand.php]: Might be better as a requires option. I think that would make the command more readable `--from=xxx --to=yyy`

Comment (cableman) [reply src/Command/ExtractStatisticsDaysCommand.php]: Would it be possible to have an default value or an special value for `--to` that is `today` or 'now' ?

File: src/Command/ExtractStatisticsDaysCommand.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        }
+
+        $rawTypes = $input->getOption('types');
+
+        $typesExploded = explode(',', $rawTypes);
+
+        $types = array_reduce($typesExploded, function ($carry, $type) {
+            if (in_array($type, ['hit', 'nohit', 'undetermined']) && !in_array($type, $carry)) {
+                $carry[] = $type;
+            }
```
Comment (cableman) [inline src/Command/ExtractStatisticsDaysCommand.php:135]: Should this have an `else` that throws an error if it contains an unknown type?

File: src/Export/CsvTarget.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+use App\Document\Entry;
+use App\Document\ExtractionResult;
+use Box\Spout\Writer\Common\Creator\WriterEntityFactory;
+
+/**
+ * Class CsvTarget.
+ */
+class CsvTarget implements ExtractionTargetInterface
+{
+    /* @var \Box\Spout\Writer\WriterInterface $writer */
```
Comment (cableman) [inline src/Export/CsvTarget.php:18]: The namespace here is not needed when it have been mentioned as use statement.

File: src/Export/CsvTarget.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        $rowFromValues = WriterEntityFactory::createRowFromArray($headers);
+        $this->writer->addRow($rowFromValues);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function finish(): void
+    {
+        $this->writer->close();
```
Comment (cableman) [inline src/Export/CsvTarget.php:75]: As an extra level you could also make a `__destruct()` function that calls this function to ensure proper file closer on errors/exceptions?

File: src/Export/CsvTarget.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        // Since this is a spreadsheet export, we are not concerned with already added entries.
+        return false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function recordExtractionResult(ExtractionResult $extractionResult): void
+    {
```
Comment (cableman) [inline src/Export/CsvTarget.php:111]: Comment about why this is empty would be [grrreat](https://upload.wikimedia.org/wikipedia/en/thumb/c/c8/Tony_the_Tiger_%28Kellogg%27s_Frosted_Flakes%27_mascot%29.jpg/220px-Tony_the_Tiger_%28Kellogg%27s_Frosted_Flakes%27_mascot%29.jpg) (tony)

File: src/Export/CsvTarget.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function flush(): void
+    {
```
Comment (cableman) [inline src/Export/CsvTarget.php:119]: Same here... missing comment.

File: src/Export/CsvTarget.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+     */
+    public function setExtractionTypes(array $types): void
+    {
+        $this->types = $types;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function acceptType(string $type): bool
```
Comment (cableman) [inline src/Export/CsvTarget.php]: Should this have been named `acceptsType` or `isTypeAccepted` ?

File: src/Export/MongoDBTarget.php [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    private $entryRepository;
+    /* @var array $types */
+    private $types;
+
+    /**
+     * MongoDBTarget constructor.
+     *
+     * @param \Doctrine\ODM\MongoDB\DocumentManager $documentManager
+     * @param \App\Repository\EntryRepository $entryRepository
+     * @param \App\Repository\ExtractionResultRepository $extractionResultRepository
```
Comment (cableman) [inline src/Export/MongoDBTarget.php]: All these namespaces should be removed to make it more readable. They are not need if the `use` statement for the namespace, well, have been used.

File: src/Export/MongoDBTarget.php [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        $this->documentManager = $documentManager;
+        $this->extractionResultRepository = $extractionResultRepository;
+        $this->entryRepository = $entryRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function initialize(): void
+    {
```
Comment (cableman) [inline src/Export/MongoDBTarget.php:44]: Would be nice with comment why these are empty in this target?

**Review by cableman — CHANGES_REQUESTED** [review]
Are there written tests for this new stuff or are they covered by the existing tests?

File: src/Command/ExtractStatisticsDaysCommand.php (hunk 4 above)
Comment (tuj) [reply src/Command/ExtractStatisticsDaysCommand.php:135]: added

File: src/Export/CsvTarget.php (hunk 6 above)
Comment (tuj) [reply src/Export/CsvTarget.php:75]: added

File: src/Export/CsvTarget.php (hunk 7 above)
Comment (tuj) [reply src/Export/CsvTarget.php:111]: added

File: src/Export/MongoDBTarget.php (hunk 11 above)
Comment (tuj) [reply src/Export/MongoDBTarget.php:44]: added

File: src/Export/CsvTarget.php (hunk 8 above)
Comment (tuj) [reply src/Export/CsvTarget.php:119]: added

**Review by cableman — APPROVED** [review]

**Pull request merged**